### PR TITLE
Convert io_service_ from reference to smart pointer

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Storing a reference causes a compile warning that's turned into an error. Using smart pointers will make sure the lifetime is handled correctly and there should be no performance impact (since shared pointers are only constructed rarely).

## Related issue number

#631

#6231

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/gcs/redis_context.h [hunk 1]
```diff
@@ -91,19 +91,19 @@ class RedisCallbackManager {
...
-        io_service_.post([self, reply]() { self->callback_(std::move(reply)); });
+        io_service_->post([self, reply]() { self->callback_(std::move(reply)); });
       }
     }
 
     RedisCallback callback_;
     bool is_subscription_;
     int64_t start_time_;
-    boost::asio::io_service &io_service_;
+    boost::asio::io_service *io_service_;
```
Comment (ericl) [inline src/ray/gcs/redis_context.h]: Actually could we use std::shared_ptr<> here instead? That should have minimal overhead (paid once at construction time), and avoids a bunch of hazards with lifetime management.

Comment (mehrdadn) [reply src/ray/gcs/redis_context.h]: Sure, if you'd like to push the change! I'm not familiar with the code here; just trying to get compile errors out of the way at the moment.

**Review by pcmoritz — APPROVED** [review]
LGTM!

**Review by raulchen — COMMENTED** [review]
Could you elaborate on what warning you saw? And why didn't the warning occur before?

The drawback of this change is that it'd make the ownership unclear.

**Comment by mehrdadn** [comment]
@raulchen It's in the PR I linked to, in this comment: https://github.com/ray-project/ray/pull/6231#discussion_r350463552

That said, I only wanted to do the rename. I agree with you that changing to `shared_ptr` makes the ownership unclear... I don't think it's a good idea either. But @pcmoritz (and @ericl) would like me to do it anyway so you'll have to convince them...

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**